br/pkg/utils/error_handling.go

```diff
@@ -0,0 +1,181 @@
+	return &ErrorContext{
+		description:              scenario,
+		encounterTimes:           make(map[uint64]int),
+		encounterTimesLimitation: limitation,
+	}
+}
+
+func NewDefaultContext() *ErrorContext {
+	return &ErrorContext{
+		description:              "default",
+		encounterTimes:           make(map[uint64]int),
+		encounterTimesLimitation: 1,
+	}
+}
+
+func HandleBackupError(err *backuppb.Error, storeId uint64, ec *ErrorContext) ErrorHandlingResult {
+	if err == nil {
+		return ErrorHandlingResult{Retry, unreachableRetryMsg}
+	}
+	res := handleBackupProtoError(err)
+	// try the best effort handle unknown error based on their error message
+	if res.Strategy == Unknown && len(err.Msg) != 0 {
+		return HandleUnknownBackupError(err.Msg, storeId, ec)
+	}
+	return res
+}
+
+func handleBackupProtoError(e *backuppb.Error) ErrorHandlingResult {
+	switch e.Detail.(type) {
+	case *backuppb.Error_KvError:
+		return ErrorHandlingResult{Retry, retryOnKvErrorMsg}
+	case *backuppb.Error_RegionError:
+		return ErrorHandlingResult{Retry, retryOnRegionErrorMsg}
+	case *backuppb.Error_ClusterIdError:
+		return ErrorHandlingResult{GiveUp, clusterIdMismatchMsg}
+	}
+	return ErrorHandlingResult{Unknown, unknownErrorMsg}
+}
+
+// HandleUnknownBackupError TODO: remove this method and map all the current unknown errors to an error type
+func HandleUnknownBackupError(msg string, uuid uint64, ec *ErrorContext) ErrorHandlingResult {
+	// UNSAFE! TODO: use meaningful error code instead of unstructured message to find failed to write error.
+	logger := log.L().With(zap.String("description", ec.description))
+	if messageIsNotFoundStorageError(msg) {
+		reason := fmt.Sprintf("File or directory not found on TiKV Node (store id: %v). "+
+			"workaround: please ensure br and tikv nodes share a same storage and the user of br and tikv has same uid.",
+			uuid)
+		return ErrorHandlingResult{GiveUp, reason}
+	}
+	if messageIsPermissionDeniedStorageError(msg) {
+		reason := fmt.Sprintf("I/O permission denied error occurs on TiKV Node(store id: %v). "+
+			"workaround: please ensure tikv has permission to read from & write to the storage.",
+			uuid)
+		return ErrorHandlingResult{GiveUp, reason}
+	}
+	msgLower := strings.ToLower(msg)
+	if strings.Contains(msgLower, contextCancelledMsg) {
+		return ErrorHandlingResult{GiveUp, contextCancelledMsg}
+	}
+
+	if MessageIsRetryableStorageError(msg) {
+		logger.Warn(retryableStorageErrorMsg, zap.String("error", msg))
+		return ErrorHandlingResult{Retry, retryableStorageErrorMsg}
+	}
+
+	// retry enough on same store
+	ec.mu.Lock()
+	defer ec.mu.Unlock()
+	ec.encounterTimes[uuid]++
+	if ec.encounterTimes[uuid] <= ec.encounterTimesLimitation {
+		return ErrorHandlingResult{Retry, retryOnUnknownErrorMsg}
+	}
+	return ErrorHandlingResult{GiveUp, noRetryOnUnknownErrorMsg}
+}
+
+// messageIsNotFoundStorageError checks whether the message returning from TiKV is "NotFound" storage I/O error
+func messageIsNotFoundStorageError(msg string) bool {
```

Comment:
@3pointer does the storage here mean S3 or any backup destination? At first I thought it's TiKV but there is a proto type for it already so got a bit confused. Thanks

Reply:
at this context storage means TiKV.